pkg/database/sql/mariadb.go

```diff
@@ -132,6 +135,45 @@ type mariaDBDatabase struct {
 	*sqlDatabaseImpl
 }
 
+func (db *mariaDBDatabase) WriteResource(ctx context.Context, k8sObj *unstructured.Unstructured, data []byte, lastUpdated time.Time) (bool, error) {
+	tx, err := db.db.BeginTx(ctx, nil)
+	if err != nil {
+		return false, fmt.Errorf("could not begin transaction for resource %s: %s", k8sObj.GetUID(), err)
+	}
+	query, args := db.inserter.ResourceInserter(
+		string(k8sObj.GetUID()),
+		k8sObj.GetAPIVersion(),
+		k8sObj.GetKind(),
+		k8sObj.GetName(),
+		k8sObj.GetNamespace(),
+		k8sObj.GetResourceVersion(),
+		lastUpdated,
+		models.OptionalTimestamp(k8sObj.GetDeletionTimestamp()),
+		data,
+	).BuildWithFlavor(db.flavor)
+	_, execErr := tx.ExecContext(
+		ctx,
+		query,
+		args...,
+	)
+	if execErr != nil {
+		rollbackErr := tx.Rollback()
+		if rollbackErr != nil {
+			return false, fmt.Errorf("write to database failed: %s and unable to roll back transaction: %s", execErr, rollbackErr)
+		}
+		return false, fmt.Errorf("write to database failed: %s", execErr)
+	}
+	execErr = tx.Commit()
+	if execErr != nil {
+		rollbackErr := tx.Rollback()
+		if rollbackErr != nil {
+			return false, fmt.Errorf("commit to database failed: %s and unable to roll back transaction: %s", execErr, rollbackErr)
+		}
+		return false, fmt.Errorf("commit to database failed and the transactions was rolled back: %s", execErr)
+	}
+	return true, nil
```

Comment:
With this new approach I would add a //FIXME comment with an issue ID in a comment to address retrieving the real information of if it was updated or inserted.

Reply:
Right, added.

Reply:
ah, right, I didn't remember that we had the "WHERE resource.cluster_updated_ts < $?" logic in place, so yes, we currently aren't differentiating between record updated and "record not updated because it came in an ordered different than the expected"

BUT, I think that regarding the metrics we could still consider the "false+noerror" as an "update" because it's an attempt to update but it's not due to a race condition, so, if we want to know how many updates we have per kind in a timeframe compared with the inserts, we could count it as an update. Also, even though the record wasn't changed, the database connection was done, so also still useful if we want to aggregate inserts+updates to know about the amount of database interactions with regards to the sink

Reply:
ah, just read your answer [here](https://github.com/kubearchive/kubearchive/issues/985#issuecomment-2792011237)

so could we use a type different from bool to have 3 possible values and be aware of when the record is neither updated or inserted?

Reply:
I introduced an enum